### DISCUSSION TITLE
Автоматическое присвоение порядкового номера

### DISCUSSION THREAD

**Discussion by Grume** [totumonline/totum-ru-issues-and-discussions#51] [Как это сделать?] [answered]
В таблице есть числовая переменная `#nomer`, которая обозначает порядковый номер строки в таблице.  Нужно при добавлении новой строки присваивать следующий номер. Как это лучше сделать?

**Comment by erpic-ru** [comment, marked as answer]
Вычислять кодом. Вот пример с комментариями автора :)

```
// Если это строка добавления, то берем максимальное значение номера и прибавляем 1.
= : if(condition: #id = ""; then: $calc2; else: $if2)

// Если предыдущее значение равно пустоте, то значит эта позиция первая.
if2: if(condition: $sel = ""; then: 1; else: $calc)

calc: $sel + 1

// Берем значение nomer из предыдущей строки в процессе вычисления только один раз ~
~sel: select(table: $#ntn; field: 'nomer_stroki'; where: 'id' < #id; order: 'id' desc)

// Берем самый большой номер и прибавляем 1.
calc2: select(table: $#ntn; field: 'nomer_stroki'; order: 'id' desc) + 1
```

**Еще по этой теме помимо этой ветки: https://github.com/totumonline/totum-ru-issues-and-discussions/discussions/103**

**Reply by EKhorikov** [reply]
А если нужно так же, присваивать порядковые номера циклам? Не в соответствии с id, а именно порядковый номер присваивать. Создали 1,2,3, удалили 3, следующий цикл, не смотря на id = 4 будет создан =3.

**Reply by erpic-ru** [reply]
Для этого можно создать отдельное поле в расчетной таблице в цикле, которое тем же кодом (только запросы к ярлыку в таблице циклов или предыдущей расчетной). Будет вычислять порядковый номер на основе предыдущего значения.

Можно, конечно, чтобы это делала сама таблица циклов и затем забирать его в расчетную. Но кажется, что первый вариант лучше.

Смысл дополнительного номера только в том, чтобы обеспечить последовательность без провалов при удалении последнего или последних циклов. В большинстве сценариев можно пренебречь и не городить доп. номера.

Но если нужно всегда поддерживать/пересчитывать нумерацию уже созданных циклов, даже при удалении из середины таблицы, то стоит добавить код действия при удалении, который будет пересчитывать все более поздние значения:

= : recalculate( table: '....' ;where 'id' > #id )

**Reply by EKhorikov** [reply]
Тут проблема вся в том, что если это учёт, например, заказов от клиентов, и они привязаны к id, то 
1. Пересчитывать ничего нельзя, т.к. порушатся связи с привязанными по id документами.
2. При увеличении кол-ва заказов до нескольких тысяч, с этим придётся только мириться, т.к. обнуление всех номеров (например с наступлением следующего расчётного года) невозможно, ведь это idшники. А хотелось бы. Год закончился, нумерация обнулилась, а idшники пошли дальше.

**Reply by erpic-ru** [reply]
Заказы – да, не нужно пересчитывать. В этом случае
добавляется скрытое поле в хедер таблицы циклов, к которому каждый раз идет обращение за номером. 
Что касается сброса нумерации с НГ, то добавить в крон задание, которое будет обнулять это поле в нужных таблицах под бой курантов.

**Reply by ttm-alexeypolunin** [reply]
Если у вас привязка по id, то даже если вы пересчитаете и у них изменятся номера то для машины все будет ок — мозг погорит у людей.

Коды в начале этого треда как раз обрабатывают ваш случай с удалением последнего. Год пропишите в where при выборе последнего.

В бухгалтерии иногда нужно, что бы не было дырок в номерах и пересчитывать километр последующих циклов нельзя — мы как-то это обсуждали и решили, что делается дополнительная простая таблица в которую пишется номер удаляемого цикла и в коде присвоения номера сначала смотрится там.

Но тут есть подстава связанная с тем, что код не сможет удалить использованный номер из этой доп таблицы — сможет пересчитать там строку тк `recalculate` работает в кодах. И там уже считается `чекбокс`, который смотрит использование.

PS: @erpic-ru в предыдущем посте дело предлагает — так реально очень просто!

**Reply by EKhorikov** [reply]
> Заказы – да, не нужно пересчитывать. В этом случае добавляется скрытое поле в хедер таблицы циклов, к которому каждый раз идет обращение за номером. Что касается сброса нумерации с НГ, то добавить в крон задание, которое будет обнулять это поле в нужных таблицах под бой курантов.

Я правильно понял, код остаётся тот-же, но вместо обращения к id, обращаемся к этому техническому полю?
Если пример в коде напишите, будет вообще отлично, я новичок, пока только разбираюсь)))

**Reply by erpic-ru** [reply]
Код будет ещё проще. Создается 2 поля.
1. Техническое поле в хэдере без кодов, тип - число
2. Номер документа в цикле, тип - число, закрыто на редактирование

Код при добавлении: 
`= : @zakazy.h_nomer + 1 (обращение к полю)`

После отработки кода выполняется 
Код действия при добавлении: 
`= : set(table: 'zakazy'; field: 'h_nomer' = $#nfv;)`

На любое поле в таблице циклов (или лучше разместить в поле ярлыка с номером цикла) можно поставить код действия при удалении: 
```
 =: if(condition: $#onfv=@zakazy.h_nomer; then: $1; )
1: set(table: 'zakazy'; field: 'nomer_zakaza' = $prev )
prev: $#onfv - 1

```
Все это будет работать, если пользователи не добавляют циклы в прошлом году, задним числом. Если такое есть, то надежней через where искать последний номер в нужном году.

**Reply by Em2see** [reply]
@erpic-ru 
Я пытаюсь найти ваш контак, так как сайт erpic.ru недоступен.
Если есть возможность напишите e2m3x4@mail.ru; или телеграм @sun_n_sky

**Reply by Grume** [reply]
<img width="110" alt="image" src="https://user-images.githubusercontent.com/24443962/161908366-f25df62b-58ec-4a8f-9bc3-c7ff134742be.png">
Когда вручную меняешь номера, тем самым нарушая связь номера с id, то вычисление номера происходит неверное, так как он вычисляет основываясь на id. Как нужно изменить, чтобы код вычислял значение не по id, а пройдясь по всем номерам находил самое большое и прибавлял туда

**Reply by Grume** [reply]
> <img alt="image" width="110" src="https://user-images.githubusercontent.com/24443962/161908366-f25df62b-58ec-4a8f-9bc3-c7ff134742be.png">
> 
> Когда вручную меняешь номера, тем самым нарушая связь номера с id, то вычисление номера происходит неверное, так как он вычисляет основываясь на id. Как нужно изменить, чтобы код вычислял значение не по id, а пройдясь по всем номерам находил самое большое и прибавлял туда

Если в настройках таблицы выставить сортировку по полю номера, то вычисления становятся корректные, вопрос закрыт

**Reply by ttm-alexeypolunin** [reply]
Чесн не уверен, что поле сортировки в этом случае решает проблему именно правильного присвоения следующего номера — отображения порядка да.

Чтоб гарантировано работало — вам нужно в коде который вычисляет номер основываться не на id, а на номерах в `order`:

```
=: select(table: $#ntn; field: 'nomer_stroki'; order: 'nomer_stroki' desc) + 1
```

**Comment by DaveMaks** [comment]
Подскажите, если у меня задача : При добовлении новой записи простая таблица Т1, паралельно в таблице  Т2 тоже нужно сделать  insert, id новой записи из Т2, положить в соседнее поле Т1 (сделать ввязь многоие к одному). я сделел через код действия поля при добавлении.
```
ad1= : set(table: $#ntn; field: 'nomer_proby'=$ID_LAB)
ID_LAB: insert(table: 'laboratory_proby'; field: 'iniciator_id' = $#nu; inserts: "inserts"; log: true)
```
При таком подходе значение вставляется в первое уже существующее поле. 
По какому пути мне пойти, тоже как то через служебное поле? или каким другим способом поменять соседнее поле в момент сохранения записи?
была идея зайти через код действия таблицы, но чтото как то сложно получается)

**Reply by ttm-alexeypolunin** [reply]
Если я правильно все понял, то вам

1. Выполняем inserts и в параметре inserts указываем переменную в которую упадут id добавленный строк

2. Делаем set в туже строку `where: 'id' = #id` взяв из переменной в которой у нас id первое значение (тк там всегда список а у нас здесь одна строка добавляется):

```
ad1=: insert(table: 'laboratory_proby'; field: 'iniciator_id' = $#nu; inserts: "new_rows_ids"; log: true)

ad2=: set(table: $#ntn; field: 'nomer_proby' = $#new_rows_ids[0]; where: 'id' = #id)
```

Такая схема достаточно часто используется!
